src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -190,14 +193,31 @@ public static IPrincipal CurrentPrincipal
         {
             get
             {
-                return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
+                if (t_principal is null)
+                {
+                   CurrentPrincipal = AppDomain.CurrentDomain.GetThreadPrincipal();
+                }
+                return t_principal;
             }
             set
             {
-                CurrentThread._principal = value;
+                if (s_asyncLocalPrincipal is null)
+                {
+                    if (value is null)
+                    {
+                        return;
+                    }
+                    Interlocked.CompareExchange(ref s_asyncLocalPrincipal, new AsyncLocal<IPrincipal>(CurrentPrincipalContextChanged), null);
+                }
+                s_asyncLocalPrincipal.Value = value;
```

Comment:
The code here is indirectly setting t_principal, but it's not obvious from the code.  Can we add an assert, something like:
```C#
Debug.Assert(value == null || t_principal != null);
```
?

Reply:
ok